### PR TITLE
Compute jumpdests using the KCFGSemantics.custom_step heuristic

### PR DESCRIPTION
**Pull request by anvacaru**
~-blocked on: https://github.com/runtimeverification/k/pull/4376~
~-blocked on: https://github.com/runtimeverification/evm-semantics/pull/2450~

Implementing a `custom_step` heuristics for KEVMSemantics to compute the jumpdests using kevm-pyk.
Adding the `#loadProgram` as a cut-rule.

### CODE REVIEW

**Comment by anvacaru** [comment]
Ran the test-pyk-prove stage with `--use-booster` active on both this branch and the `v1.0.563`. 

```
| Test                                           | pyjumpdests time | master time | (pyjumpdests/master) time 
|------------------------------------------------|------------------|-------------|---------------------------
| benchmarks/requires01-a0gt0-spec.k             | 37.93            | 42.85       | 0.8851808634772462        
| benchmarks/overflow00-nooverflow-spec.k        | 35.06            | 38.7        | 0.9059431524547804        
| mcd/vow-fess-fail-rough-spec.k                 | 110.27           | 117.9       | 0.9352841391009329        
| bihu/forwardToHotWallet-failure-3-spec.k       | 64.24            | 68.4        | 0.9391812865497075        
| functional/evm-int-simplifications-spec.k      | 76.01            | 80.9        | 0.9395550061804697        
| mcd/vat-move-diff-rough-spec.k                 | 95.97            | 99.87       | 0.9609492340042054        
| functional/int-simplifications-spec.k          | 109.21           | 113.4       | 0.9630511463844796        
| functional/abi-spec.k                          | 84.91            | 87.97       | 0.9652154143457997        
| benchmarks/ecrecover00-siginvalid-spec.k       | 50.68            | 48.91       | 1.0361889184215907        
| benchmarks/ecrecoverloop00-sig0-invalid-spec.k | 54.91            | 52.56       | 1.044710806697108         
| erc20/hkg/totalSupply-spec.k                   | 66.14            | 62.45       | 1.0590872698158527        
| bihu/forwardToHotWallet-success-1-spec.k       | 46.47            | 43.24       | 1.0746993524514339        
| benchmarks/ecrecoverloop00-sigs-valid-spec.k   | 70.3             | 65.28       | 1.0768995098039216        
| benchmarks/ecrecoverloop00-sig1-invalid-spec.k | 103.45           | 94.2        | 1.0981953290870488        
| benchmarks/staticarray00-spec.k                | 39.26            | 34.79       | 1.1284851968956597        
| benchmarks/ecrecover00-sigvalid-spec.k         | 50.91            | 44.83       | 1.1356234664287308        
| TOTAL                                          | 1095.72          | 1096.25     | 0.9995165336374002        

```

File: kevm-pyk/src/kevm_pyk/kevm.py
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kevm.py:157]:
```suggestion
        :return: If the K_CELL matches the load_pattern, a Step with depth 1 is returned together with the new configuration, also registering that the `EVM.program.load` rule has been applied. Otherwise, None is returned.
```

File: kevm-pyk/src/tests/unit/test_kevm.py
Comment (palinatolmach) [inline kevm-pyk/src/tests/unit/test_kevm.py:57]: Since `BYTES` tokens are handled as a separate case, maybe it'd be safer to add a test for it? E.g.,

```
b"`\x80`@R4\x80\x15a\x00\x10W`\x00\x80\xfd[P`@Qa\x01a8\x03\x80a\x01a\x839\x81\x01`@\x81\x90Ra\x00/\x91a\x00KV[Pa\x01\x14V[cNH{q`\xe0\x1b`\x00R`A`\x04R`$`\x00\xfd[`\x00` \x80\x83\x85\x03\x12\x15a\x00^W`\x00\x80\xfd[\x82Q`\x01`\x01`@\x1b\x03\x80\x82\x11\x15a\x00uW`\x00\x80\xfd[\x81\x85\x01\x91P\x85`\x1f\x83\x01\x12a\x00\x89W`\x00\x80\xfd[\x81Q\x81\x81\x11\x15a\x00\x9bWa\x00\x9ba\x005V[`@Q`\x1f\x82\x01`\x1f\x19\x90\x81\x16`?\x01\x16\x81\x01\x90\x83\x82\x11\x81\x83\x10\x17\x15a\x00\xc3Wa\x00\xc3a\x005V[\x81`@R\x82\x81R\x88\x86\x84\x87\x01\x01\x11\x15a\x00\xdbW`\x00\x80\xfd[`\x00\x93P[\x82\x84\x10\x15a\x00\xfdW\x84\x84\x01\x86\x01Q\x81\x85\x01\x87\x01R\x92\x85\x01\x92a\x00\xe0V[`\x00\x86\x84\x83\x01\x01R\x80\x96PPPPPPP\x92\x91PPV[`?\x80a\x01\"`\x009`\x00\xf3\xfe`\x80`@R`\x00\x80\xfd\xfe\xa2dipfsX\"\x12 \xce\x0fs\x04\xc1\x8d\x10\xc8\xefe\x12(+\xa0\xff(\xa2\x95W\xa4\xeb\r\xc2\xf0\xbb((\x83\xfe\xed\t dsolcC\x00\x08\x19\x003\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@" +Bytes VV1_b_114b9705:Bytes
```

That corresponds to deploying 
```solidity
contract Constr {
    constructor(bytes memory b) {  }
}
```

**Review by palinatolmach — COMMENTED** [review]
Looks good to me! I just left one test suggestion, but would defer to Petar's review for the final approval. 

Do you think it might make sense to check if all Kontrol tests are still passing with this PR before merging? If you've already done it, please nvm.

File: kevm-pyk/src/tests/unit/test_kevm.py
Comment (palinatolmach) [reply kevm-pyk/src/tests/unit/test_kevm.py:57]: After a discussion with Andrei, I take this comment back — `bytes` constructor arguments aren't supported and are problematic because they have symbolic length and, therefore, offsets. They don't seem common enough to get it addressed now.

**Comment by anvacaru** [comment]
> Do you think it might make sense to check if all Kontrol tests are still passing with this PR before merging?

I ran the kontrol test suite, and there is no proof getting stuck. However, on the CSE tests it looks like some of the summaries are no longer getting applied and I need to investigate that.

File: kevm-pyk/src/kevm_pyk/kevm.py
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kevm.py:671]:
```suggestion
    :param sections: A section is expected to be either a concrete sequence of bytes (Bytes) or a symbolic buffer of concrete width (#buf(WIDTH, _)).
```

File: kevm-pyk/src/kevm_pyk/kevm.py
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kevm.py:681]: Could we also assert here that the token is of sort `Int` or does that hold by construction?

File: kevm-pyk/src/tests/unit/test_kevm.py
Comment (PetarMax) [inline kevm-pyk/src/tests/unit/test_kevm.py:57]: Could we also please have a test in which we have concrete bytecode interspersed with symbolic buffers?

**Review by PetarMax — COMMENTED** [review]
Left several comments, requesting one more test.

File: kevm-pyk/src/tests/unit/test_kevm.py
Comment (anvacaru) [reply kevm-pyk/src/tests/unit/test_kevm.py:57]: Added more tests.

File: kevm-pyk/src/kevm_pyk/kevm.py
Comment (anvacaru) [reply kevm-pyk/src/kevm_pyk/kevm.py:680]: I'd say that it holds by construction of `#buf`. I'm not sure how possible is to have a symbolic length, but in this case, `args[0]` would be a `KVariable` instead of a `KToken`.

**Comment by PetarMax** [comment]
LGTM. Let's understand what is going wrong with the Kontrol tests and then we can merge.

File: kevm-pyk/src/kevm_pyk/kevm.py
Comment (tothtamas28) [inline kevm-pyk/src/kevm_pyk/kevm.py:153]: Based on your measurements, this does not seem to improve performance. What's the purpose of this change?

Comment (anvacaru) [reply kevm-pyk/src/kevm_pyk/kevm.py:153]: At this moment we cannot compute jump destinations for bytecodes that have symbolic values, blocking issues such as https://github.com/runtimeverification/kontrol/pull/573 and https://github.com/runtimeverification/evm-semantics/issues/2219.

Comment (tothtamas28) [reply kevm-pyk/src/kevm_pyk/kevm.py:153]: I see, thank you. Please include these details in the PR description.

File: kevm-pyk/src/tests/unit/test_kevm.py
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test_kevm.py:69]:
```suggestion
            KEVM.buf(token(32), KVariable('VV0_x', 'Int')),
```

Comment (tothtamas28) [reply kevm-pyk/src/tests/unit/test_kevm.py:69]: Opened an issue that's related: #2456

File: kevm-pyk/src/tests/unit/test_kevm.py
Comment (tothtamas28) [inline kevm-pyk/src/tests/unit/test_kevm.py:95]: I think

```suggestion
            token(b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00['),
```

should work in cases like this.

Comment (tothtamas28) [reply kevm-pyk/src/tests/unit/test_kevm.py:95]: Indeed:

```python3
>>> from pyk.kast.inner import KToken
>>> from pyk.prelude.utils import token
>>> token(b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00[') == KToken('b"\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00["', 'Bytes')
True
```

Comment (anvacaru) [reply kevm-pyk/src/tests/unit/test_kevm.py:95]: Nice! I didn't know `token` was a thing! I'll use it from now on instead.

**Review by PetarMax — APPROVED** [review]

**Pull request merged**
